File: python-flask-reactjs-map/geo_clustering_point_merger.py

from geo_types import GeoPoint

from utils import enter, exit, debug


class _MergedClusteringPoint:

    def __init__(self, from_point, to_point):
        self.count = from_point.count + to_point.count

        self.from_point = from_point
        self.to_point = to_point

        # Compute mid point weighted for count
        self.mid_point = _MergedClusteringPoint._compute_mid_point(
            from_point, to_point)

    def get_point(self):
        return self.mid_point

    def __str__(self):
        return self.get_point().__str__()

    def __repr__(self):
        return self.__str__()

    @staticmethod
    def _compute_mid_point(from_point, to_point):

        fp = from_point.get_point()
        tp = to_point.get_point()

        # Multiply by count for weighting
        from_latitude = fp.latitude * from_point.count
        from_longitude = fp.longitude * from_point.count

        to_latitude = tp.latitude * to_point.count
        to_longitude = tp.longitude * to_point.count

        sum_count = from_point.count + to_point.count

        computed_latitude = (from_latitude + to_latitude) / sum_count
        computed_longitude = (from_longitude + to_longitude) / sum_count

        return GeoPoint(computed_latitude, computed_longitude)


'''

Helper class for merging a series of points from distances below a certain max

'''


class GeoClusteringPointMerger:

    @staticmethod
    def merge_points_with_distances_below_max(indent, distances, max_diameter_km):

        enter(indent, 'GeoClusteringPointMerger.merge_points_with_distances_below_max',
              'distances=' + str(distances.count()))

        # List of only the points that have shorter distance than max
        distances_below_max = distances.find_below_max(max_diameter_km)

        debug(indent, 'GeoClusteringPointMerger.merge_points_with_distances_below_max', 'Got distances below max ' +
              str(distances_below_max.count()))

        # Merge those distance points, and return a set of removed points from merged distances
        merged_points, removed_points = GeoClusteringPointMerger._merge_distances(
            distances_below_max)

        debug(indent, 'GeoClusteringPointMerger.merge_points_with_distances_below_max', 'after merging got merged ' + str(len(merged_points)) + ', removed ' +
              str(len(removed_points)))

        updated_distances = distances.remove_distances_with_points(
            removed_points)

        debug(indent, 'GeoClusteringPointMerger.merge_points_with_distances_below_max',
              'removing points from ' + str(distances.count()) + ' to ' + str(updated_distances.count()))

        exit(indent, 'GeoClusteringPointMerger.merge_points_with_distances_below_max',
             'returning ' + str(len(merged_points)) + ' from ' + str(distances_below_max.count()))

        return merged_points, updated_distances

    @staticmethod
    def _merge_distances(distances, debug=False):

        distances_set = distances.to_set()
        merged_points = []
        removed_points = set()

        while len(distances_set) != 0:

            to_iterate = list(distances_set)

            for distance in to_iterate:
                # Can merge points unless one of them were already removed
                if distance.from_point in removed_points or distance.to_point in removed_points:
                    if debug:
                        print 'Already removed'

                    distances_set.remove(distance)
                else:
                    # Merge point and remove
                    removed_points.add(distance.from_point)
                    removed_points.add(distance.to_point)

                    distances_set.remove(distance)

                    merged = _MergedClusteringPoint(
                        distance.from_point,
                        distance.to_point)

                    merged_points.append(merged)

        if len(distances_set) != 0:
            raise Exception('Expected empty set')

        return merged_points, removed_points

File: python-flask-reactjs-map/geo_clustering/c/geo_clustering_point_merger.c
#include "debug.h"

#include "geo_clustering_point_merger.h"
#include "geo_clustering_types.h"

static void merged_clustering_point_init(
        indent_t indent,
        geo_merged_point_t *const point,
        const geo_scratch_distance_t *const to_merge);

static void compute_mid_point(
        indent_t indent,
        geo_merged_point_t *const dst,
        const geo_clustered_point_t *const from,
        const geo_clustered_point_t *const to);


int32_t merge_distances_below_max(
    indent_t indent,
    const geo_scratch_distance_t *const distances,
    uint32_t num_distances,
    float max_diameter_km,
    scratch_buf_t *out_merged_points,
    bitmap_t *out_removed_points
) {

    boolean ok = TRUE;

    uint32_t merged_points_added = 0;

    enter(indent, "num_distances=%d, max_diameter_km=%f", num_distances, max_diameter_km);

    for (int i = 0; i < num_distances; ++ i) {

        const geo_scratch_distance_t *const distance = &distances[i];

        debug(indent, "considering point at index %d from (%f, %f) index %d to (%f, %f) index %d",
                    i,
                    distance->from_point.geo_point.latitude,
                    distance->from_point.geo_point.longitude,
                    distance->from_point_index,
                    distance->to_point.geo_point.latitude,
                    distance->to_point.geo_point.longitude,
                    distance->to_point_index);

        debug(indent, "bitmap initial byte is 0x%01x", UB(out_removed_points)[0]);
   
        debug(indent, "is_set %d %d",
                bitmap_is_set(out_removed_points, distance->from_point_index),
                bitmap_is_set(out_removed_points, distance->to_point_index));


     
        if (    bitmap_is_set(out_removed_points, distance->from_point_index)
             || bitmap_is_set(out_removed_points, distance->to_point_index)) {

            /* Already merged one or both points, so distance no longer usable */
            debug_(indent, "skipping distance already set");
        }
        else {

            /* Check that distance within max diameter, if so ignore */
            if (distance->distance > max_diameter_km) {
                debug_(indent, "skipping distance due to outside of range");
                continue;
            }

            /* Can merge the distance since none of points already merged */
            bitmap_set(out_removed_points, distance->from_point_index);
            bitmap_set(out_removed_points, distance->to_point_index);

            /* Add merged point */
            if (merged_points_added == out_merged_points->nmemb) {
                /* Must allocate more space in buffer */
                if (!scratch_buf_realloc(out_merged_points)) {
                    ok = FALSE;
                    break;
                }
            }

            geo_merged_point_t *dst = scratch_buf_at(out_merged_points, merged_points_added ++);

            merged_clustering_point_init(indent + 1, dst, distance);
        }
    } 

    const int32_t result = ok ? merged_points_added : -1;

    exit(indent, "result=%d", result);

    return result;
}


static void merged_clustering_point_init(
        indent_t indent,
        geo_merged_point_t *const point,
        const geo_scratch_distance_t *const to_merge) {


   compute_mid_point(
        indent,
        point,
        (const geo_clustered_point_t *)&to_merge->from_point,
        (const geo_clustered_point_t *)&to_merge->to_point); 

}

static void compute_mid_point(
        indent_t indent,
        geo_merged_point_t *const dst,
        const geo_clustered_point_t *const from,
        const geo_clustered_point_t *const to) {

    const float from_latitude = from->geo_point.latitude * from->count;
    const float from_longitude = from->geo_point.longitude * from->count; 

    const float to_latitude = to->geo_point.latitude * to->count;
    const float to_longitude = to->geo_point.longitude * to->count;

    const uint32_t sum_count = from->count + to->count;

    const float computed_latitude = (from_latitude + to_latitude) / sum_count;
    const float computed_longitude = (from_longitude + to_longitude) / sum_count;

    enter(indent, "from=(%f, %f) count %d, to=(%f, %f) count %d",
            from->geo_point.latitude,
            from->geo_point.longitude,
            from->count,
            to->geo_point.latitude,
            to->geo_point.longitude,
            to->count);

    debug(indent, "set sum count %d from %d and %d",
                sum_count,
                from->count,
                to->count);

    dst->count = sum_count;
    dst->geo_point.latitude = computed_latitude;
    dst->geo_point.longitude = computed_longitude;

    exit(indent, "dst=(%f, %f) count=%d",
        dst->geo_point.latitude,
        dst->geo_point.longitude,
        dst->count);

}




